app/src/test/java/net/bonysoft/magicmirror/facerecognition/KeyboardFaceSourceTest.java

```diff
@@ -0,0 +1,84 @@
+package net.bonysoft.magicmirror.facerecognition;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.mockito.internal.verification.VerificationModeFactory.times;
+
+@RunWith(MockitoJUnitRunner.class)
+public class KeyboardFaceSourceTest {
+
+    private static final int CODE_SAD = 1;
+    private static final int CODE_HAPPY = 2;
+
+    private KeyboardFaceSource source;
+    @Mock
+    private FaceTracker.FaceListener mockListener;
+    @Mock
+    private KeyToFaceMappings mockMappings;
+
+    @Before
+    public void setUp() throws Exception {
+        source = new KeyboardFaceSource(mockListener, mockMappings);
+        when(mockMappings.getFace(CODE_SAD)).thenReturn(FaceExpression.SAD);
+        when(mockMappings.getFace(CODE_HAPPY)).thenReturn(FaceExpression.HAPPY);
+
+        when(mockMappings.contains(CODE_HAPPY)).thenReturn(true);
+        when(mockMappings.contains(CODE_SAD)).thenReturn(true);
+    }
+
+    @Test
+    public void givenKeySadIsPressed_thenSadExpressionIsTriggered() {
+        source.onKeyDown(CODE_SAD);
+        verify(mockListener).onNewFace(FaceExpression.SAD);
+
+    }
+
+    @Test
+    public void givenKeySadIsPressedTwice_thenSadExpressionIsTriggeredOnlyOnce() {
+        // we are testing when user is keep the button pressed
```

Comment:
Why then not extracting the following 2 lines into a method? `keepsPressing(CODE_SAD)`